File: TOOLS/lua/test-hooks.lua
local utils = require("mp.utils")

function hardsleep()
    os.execute("sleep 1s")
end

local hooks = {"on_before_start_file", "on_load", "on_load_fail",
               "on_preloaded", "on_unload", "on_after_end_file"}

for _, name in ipairs(hooks) do
    mp.add_hook(name, 0, function()
        print("--- hook: " .. name)
        hardsleep()
        print("    ... continue")
    end)
end

local events = {"start-file", "end-file", "file-loaded", "seek",
                "playback-restart", "idle", "shutdown"}
for _, name in ipairs(events) do
    mp.register_event(name, function()
        print("--- event: " .. name)
    end)
end

local props = {"path", "metadata"}
for _, name in ipairs(props) do
    mp.observe_property(name, "native", function(prop, val)
        print("property '" .. prop .. "' changed to '" ..
              utils.to_string(val) .. "'")
    end)
end

File: TOOLS/python/test-hooks.py
import time
from mpvclient import mpv  # type: ignore

def hardsleep():
    # os.execute("sleep 1s")
    time.sleep(1)


hooks = ["on_before_start_file", "on_load", "on_load_fail",
               "on_preloaded", "on_unload", "on_after_end_file"]


def add_hook(name, priority):
    @mpv.add_hook(name, priority)
    def func():
        print("--- hook: " + name)
        hardsleep()
        print("    ... continue")


for name in hooks:
    add_hook(name, 0)

events = [
    mpv.MPV_EVENT_START_FILE,
    mpv.MPV_EVENT_END_FILE,
    mpv.MPV_EVENT_FILE_LOADED,
    mpv.MPV_EVENT_SEEK,
    mpv.MPV_EVENT_PLAYBACK_RESTART,
    mpv.MPV_EVENT_SHUTDOWN,
]


def register_to_event(name):
    @mpv.register_event(name)
    def func(event):
        print(f"--- event: {name}")


for name in events:
    register_to_event(name)


def observe(name):
    @mpv.observe_property(name, mpv.MPV_FORMAT_NODE)
    def func(value):
        print(f"property '{name}' change to '{value}'")


props = ["path", "metadata"]
for name in props:
    observe(name)
